File: src/http-client-mock/MockedHttpClientBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using JetBrains.Annotations;

namespace HttpClientMock
{
    public class MockedHttpClientBuilder
    {
        private readonly List<RequestBehaviorBuilder> builders = new List<RequestBehaviorBuilder>();

        [PublicAPI]
        public HttpClient Build(string baseAddress)
        {
            var handler = new MockHandler(builders.Select(b => b.Build()).ToArray());
            return new HttpClient(handler)
            {
                BaseAddress = new Uri(baseAddress)
            };
        }

        public RequestBehaviorBuilder WhenGet(string uri)
        {
            return WhenGet(UrlMatchers.Is(uri));
        }

        [PublicAPI]
        public RequestBehaviorBuilder WhenGet(Func<string, bool> urlMatcher)
        {
            return CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Get);
        }

        public RequestBehaviorBuilder WhenPost(string uri)
        {
            return WhenPost(UrlMatchers.Is(uri));
        }

        [PublicAPI]
        public RequestBehaviorBuilder WhenPost(Func<string, bool> urlMatcher)
        {
            return CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Post);
        }

        public RequestBehaviorBuilder WhenPut(string uri)
        {
            return WhenPut(UrlMatchers.Is(uri));
        }

        [PublicAPI]
        public RequestBehaviorBuilder WhenPut(Func<string, bool> urlMatcher)
        {
            return CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Put);
        }

        public RequestBehaviorBuilder WhenDelete(string uri)
        {
            return WhenDelete(UrlMatchers.Is(uri));
        }

        [PublicAPI]
        public RequestBehaviorBuilder WhenDelete(Func<string, bool> urlMatcher)
        {
            return CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Delete);
        }

        [PublicAPI]
        public RequestBehaviorBuilder When(Func<string, bool> urlMatcher, HttpMethod httpMethod)
        {
            return CreateRequestBehaviorBuilder(urlMatcher, httpMethod);
        }

        private RequestBehaviorBuilder CreateRequestBehaviorBuilder(Func<string, bool> urlMatcher, HttpMethod httpMethod)
        {
            var builder = new RequestBehaviorBuilder(urlMatcher, httpMethod);
            builders.Add(builder);
            return builder;
        }
    }
}

File: src/http-client-mock/MockedHttpClientBuilder.fs
namespace HttpClientMock

open System
open System.Net.Http
open JetBrains.Annotations

type MockedHttpClientBuilder() =
    
    let mutable builders: RequestBehaviorBuilder list = List.empty
    
    let CreateRequestBehaviorBuilder(urlMatcher: Func<string, bool>, httpMethod: HttpMethod) =
        RequestBehaviorBuilder(FuncConvert.FromFunc(urlMatcher), httpMethod)
            |> fun b -> builders <- builders @ [b]; b
    
    [<PublicAPI>]
    member this.Build(baseAddress: string) =
        new MockHandler(builders |> Seq.map(fun b -> b.Build() ) |> Array.ofSeq)
            |> fun h -> new HttpClient(h, BaseAddress = Uri(baseAddress))

    [<PublicAPI>]
    member this.WhenGet(urlMatcher: Func<string, bool>) =
        CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Get);

    [<PublicAPI>]            
    member this.WhenGet(url: string) =
        this.WhenGet(UrlMatchers.Is(url));
        
        
    [<PublicAPI>]
    member this.WhenPost(urlMatcher: Func<string, bool>) =
        CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Post)
    
    [<PublicAPI>]
    member this.WhenPost(uri: string) =
        this.WhenPost(UrlMatchers.Is(uri))

    [<PublicAPI>]
    member this.WhenPut(urlMatcher: Func<string, bool>) =
        CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Put)
        
    [<PublicAPI>]
    member this.WhenPut(url: string) =
        this.WhenPut(UrlMatchers.Is(url))

    [<PublicAPI>]
    member this.WhenDelete(urlMatcher: Func<string, bool>) =
        CreateRequestBehaviorBuilder(urlMatcher, HttpMethod.Delete)
        
    [<PublicAPI>]
    member this.WhenDelete(url: string) =
        this.WhenDelete(UrlMatchers.Is(url))
    
    [<PublicAPI>]
    member this.When(urlMatcher: Func<string, bool>,  httpMethod: HttpMethod) =
        CreateRequestBehaviorBuilder(urlMatcher, httpMethod);
    